internal/mining/mining_view.go

```diff
@@ -0,0 +1,464 @@
+// Copyright (c) 2020 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mining
+
+import "github.com/decred/dcrd/chaincfg/chainhash"
+
+const (
+	// AncestorTrackingLimit is the maximum number of ancestors a transaction
+	// can have ancestor stats calculated for.
+	AncestorTrackingLimit = 25
+)
+
+// defaultAncestorStats is the default-state of ancestor stats for a transaction
+// that has no ancestors.
+var defaultAncestorStats = &TxAncestorStats{}
+
+// TxMiningView represents a snapshot of all transactions in mempool
+// as well as the hierarchy of transactions, as they may spend from one another
+// in the mempool.
+type TxMiningView struct {
```

Comment:
For RPC server specifically, I think it would be mocking at a higher level (such as mocking the functions in `mempool` or `mining` that use these types internally).  If RPC server did need an interface for these types for any reason though, I think it would make sense for that interface to be defined in RPC server for its specific needs. 

It probably does make sense to have an interface for the mining view in `mempool`, but it would be a pretty big interface currently since the mempool tests are using many things from the mining view to ensure that the view is working properly.  If those tests are moved to `mining`, and then the `mempool` tests just validate that transactions are being added and removed to the mining view appropraitely, then there could be a simple interface defined there.

Reply:
Sounds good.  I did notice that the majority of the view testing is still done in `mempool` instead of in `mining` where it really would better belong now, but I figured that was primarily because you're still working on the mining test harness and at least part of that will be needed to avoid a bunch of duplicate setup code for the mining view tests as well.

Ultimately, I would definitely like to see the view tests alongside the view code in `mining`.  At that point, a small mining view interface of probably just `AddTransaction` and `RemoveTransaction` should work out for the `mempool` and, by extension, the RPC server tests nicely.

Reply:
Agreed, that is exactly what I was thinking, and will move them over in a subsequent PR.  I didn't move them yet since as you mentioned, there is a bunch of test setup code needed, and I also wanted to ensure that the existing tests still pass as things were moved.